src/NuGetForUnity/Editor/NugetPackageInstaller.cs

```diff
@@ -170,31 +171,72 @@ private static bool Install(INugetPackage package, bool refreshAssets, bool inst
                     using (var zip = ZipFile.OpenRead(cachedPackagePath))
                     {
+                        var libs = new Dictionary<string, List<ZipArchiveEntry>>();
+
                         foreach (var entry in zip.Entries)
                         {
-                            var filePath = Path.Combine(baseDirectory, entry.FullName);
+                            var entryFullName = entry.FullName;
+                            if (PackageContentManager.ShouldSkipUnpackingOnPath(entryFullName, package.Id))
+                            {
+                                continue;
+                            }
 
-                            var directory = Path.GetDirectoryName(filePath) ??
-                                            throw new InvalidOperationException($"Failed to get directory name of '{filePath}'");
-                            Directory.CreateDirectory(directory);
-                            if (Directory.Exists(filePath))
+                            // we don't want to unpack all lib folders and then delete all but one; rather we will decide the best
+                            // target framework before unpacking, but first we need to collect all lib entries from zip
+                            if (entryFullName.StartsWith("lib"))
                             {
-                                Debug.LogWarning($"The path {filePath} refers to an existing directory. Overwriting it may lead to data loss.");
+                                const int frameworkStartIndex = 4; // length of "lib/"
+                                var secondSlashIndex = entryFullName.IndexOf("/", frameworkStartIndex, StringComparison.Ordinal);
+                                var framework = entryFullName.Substring(frameworkStartIndex, secondSlashIndex - frameworkStartIndex);
+                                if (libs.TryGetValue(framework, out var entryList))
+                                {
+                                    entryList.Add(entry);
+                                }
+                                else
+                                {
+                                    libs[framework] = new List<ZipArchiveEntry> { entry };
+                                }
+
                                 continue;
                             }
 
-                            entry.ExtractToFile(filePath, true);
+                            ExtractPackageEntry(entry, baseDirectory);
+                        }
 
-                            if (ConfigurationManager.NugetConfigFile.ReadOnlyPackageFiles)
+                        // go through all lib zip entries and find the best target framework, then unpack it
+                        var bestFramework = TargetFrameworkResolver.TryGetBestTargetFramework(libs.Keys, key => key);
+                        if (bestFramework != null)
+                        {
+                            foreach (var entry in libs[bestFramework])
```

Comment:
There's a problem with this. I added a constraint to `TryGetBestTargetFramework` method that `T` has to be a class, because at first I tried what you suggested, but the default value for key-value pair is not a `null` value, rather an empty KVP ([,]), and we check for `null` there when asking for `bestMatch`. I could maybe instead ask something like `if (bestMatch != default(T))`?

Reply:
I used `if (bestFrameworkEntries.Values != null)` (see in commit)